### ISSUE TITLE
TB running flake8 with flake8-type-checking installed

### ISSUE THREAD

**Issue by ben-alkov** [snok/flake8-type-checking#30]
Factoids:

- Fedora 33
- Python 3.9.2
- flake8 3.8.4
- `flake8 --version` (non-apology; I have a linter fetish):

  ```
  3.8.4 (assertive: 1.3.0, dlint: 0.11.0, flake-mutable: 1.2.0, flake8-bandit: 2.1.2, flake8-bugbear: 20.11.1,
  flake8-cognitive-complexity: 0.1.0, flake8-comprehensions: 3.3.1, flake8-eradicate: 1.0.0,
  flake8-expression-complexity: 0.0.9, flake8-fixme: 1.1.1, flake8-mock: 0.3, flake8-pie: 0.7.1,
  flake8-pytest-style: 1.4.1, flake8-return: 1.1.2, flake8-type-checking: 1.0.0, flake8_builtins: 1.5.2,
  flake8_multiline_containers: 0.0.18, flake8_simplify: 0.14.0, flake8_typing_imports: 1.10.1,
  mccabe: 0.6.1, pycodestyle: 2.6.0, pyflakes: 2.2.0) CPython 3.9.2 on Linux
  ```

Let me know if more info is needed.

```sh
⌜10:22⌟ 🐚 flake8
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/lib64/python3.9/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/usr/lib64/python3.9/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/home/$USER/.local/lib/python3.9/site-packages/flake8/checker.py", line 676, in _run_checks
    return checker.run_checks()
  File "/home/$USER/.local/lib/python3.9/site-packages/flake8/checker.py", line 589, in run_checks
    self.run_ast_checks()
  File "/home/$USER/.local/lib/python3.9/site-packages/flake8/checker.py", line 494, in run_ast_checks
    for (line_number, offset, text, _) in runner:
  File "/home/$USER/.local/lib/python3.9/site-packages/flake8_type_checking/plugin.py", line 37, in run
    visitor = TypingOnlyImportsChecker(self._tree)
  File "/home/$USER/.local/lib/python3.9/site-packages/flake8_type_checking/checker.py", line 348, in __init__
    self.visitor.visit(node)
  File "/usr/lib64/python3.9/ast.py", line 407, in visit
    return visitor(node)
  File "/usr/lib64/python3.9/ast.py", line 483, in generic_visit
    value = self.visit(value)
  File "/usr/lib64/python3.9/ast.py", line 407, in visit
    return visitor(node)
  File "/home/$USER/.local/lib/python3.9/site-packages/flake8_type_checking/checker.py", line 332, in visit_FunctionDef
    self.generic_visit(node)
  File "/usr/lib64/python3.9/ast.py", line 483, in generic_visit
    value = self.visit(value)
  File "/usr/lib64/python3.9/ast.py", line 407, in visit
    return visitor(node)
  File "/usr/lib64/python3.9/ast.py", line 483, in generic_visit
    value = self.visit(value)
  File "/usr/lib64/python3.9/ast.py", line 407, in visit
    return visitor(node)
  File "/usr/lib64/python3.9/ast.py", line 483, in generic_visit
    value = self.visit(value)
  File "/usr/lib64/python3.9/ast.py", line 407, in visit
    return visitor(node)
  File "/usr/lib64/python3.9/ast.py", line 492, in generic_visit
    new_node = self.visit(old_value)
  File "/usr/lib64/python3.9/ast.py", line 407, in visit
    return visitor(node)
  File "/usr/lib64/python3.9/ast.py", line 492, in generic_visit
    new_node = self.visit(old_value)
  File "/usr/lib64/python3.9/ast.py", line 407, in visit
    return visitor(node)
  File "/home/$USER/.local/lib/python3.9/site-packages/flake8_type_checking/checker.py", line 291, in visit_Attribute
    node = self._set_child_node_attribute(node, ATTRIBUTE_PROPERTY, node.attr)
  File "/home/$USER/.local/lib/python3.9/site-packages/flake8_type_checking/checker.py", line 275, in _set_child_node_attribute
    setattr(node.__dict__[key], attr, val)
AttributeError: 'NoneType' object has no attribute 'flake8-type-checking_parent'
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/$USER/.local/bin/flake8", line 8, in <module>
    sys.exit(main())
  File "/home/$USER/.local/lib/python3.9/site-packages/flake8/main/cli.py", line 22, in main
    app.run(argv)
  File "/home/$USER/.local/lib/python3.9/site-packages/flake8/main/application.py", line 363, in run
    self._run(argv)
  File "/home/$USER/.local/lib/python3.9/site-packages/flake8/main/application.py", line 351, in _run
    self.run_checks()
  File "/home/$USER/.local/lib/python3.9/site-packages/flake8/main/application.py", line 264, in run_checks
    self.file_checker_manager.run()
  File "/home/$USER/.local/lib/python3.9/site-packages/flake8/checker.py", line 321, in run
    self.run_parallel()
  File "/home/$USER/.local/lib/python3.9/site-packages/flake8/checker.py", line 287, in run_parallel
    for ret in pool_map:
  File "/usr/lib64/python3.9/multiprocessing/pool.py", line 448, in <genexpr>
    return (item for chunk in result for item in chunk)
  File "/usr/lib64/python3.9/multiprocessing/pool.py", line 870, in next
    raise value
AttributeError: 'NoneType' object has no attribute 'flake8-type-checking_parent'
```

**Comment by sondrelg** [comment]
Looking at the code that triggered the error my guess would be that there is some type of python code that we haven't accounted for. Could you try to figure out what triggers the error? Without a reproducible example it'll be hard to proceed 🤔

**Comment by sondrelg** [comment]
@ben-alkov have you had a chance to look at this? Let me know if you run into any issues.

**Comment by sondrelg** [comment]
Closing this for now. Please let me know if you're able to narrow it down in any way 👍

**Comment by ben-alkov** [comment]
Apologies, I'm usually inundated with emails from GitHub; in this case I received none...

**Comment by ben-alkov** [comment]
> Thanks for the report! Did this happen the first time you ran it?

Exactly.

I've since installed Fedora 34, which currently has Python 3.9.4 (same flake8 version), and I'm seeing the same issue. I'll try poking it a little to see if I can narrow it down some.

**Comment by ben-alkov** [comment]
So, a SUPER-naive patch like

```diff
#  @ flake8_type_checking/checker.py:275 @ class ImportVisitor(ast.NodeTransformer):
          # Set the parent attribute on the current node children
          for key, value in node.__dict__.items():
              if type(value) not in [int, str, list, bool]:
-                  setattr(node.__dict__[key], attr, val)
+                  try:
+                      setattr(node.__dict__[key], attr, val)
+                  except AttributeError:
+                      continue
          return node

      def visit_Attribute(self, node: ast.Attribute) -> Any:
```

seems to let it work as expected[^1], but I reeeally don't understand any of the implications of that.

[^1]: In fact I see hits in one of my repos, e.g. "TC002: Move third-party import 'typing.Sequence' into a type-checking block" :rocket:

**Comment by ben-alkov** [comment]
Note also that (IIUC) `hasattr()` *might* be an alternative to trapping the exception.

**Comment by sondrelg** [comment]
Thanks for getting back to me about this @ben-alkov!

Any idea what python code in your project triggers the error? Your solution definitely might be the correct answer here, but without knowing exactly what python syntax triggered it I guess we can't be sure how to most appropriately address it 🙂 

If you have some more spare time, you might be able to narrow it down to a single file by adding a print statement to your flake8 code: 

If you simply add a `print(self.filename)` to your `venv/lib/python3.9/site-packages/flake8/checker.py`, in the `run_ast_checks` function on line 475, and run flake8 again, it should spit out the filenames it's handling while it's running (you could log with -vv or -vvv as well I think, but I find that harder to parse). From there, it should be possible to narrow it down further by commenting out parts of the code in the file you found cause the error.

Alternatively, if any of your code is public, I can take a look myself! 👏

Thanks again 🙂

**Comment by ben-alkov** [comment]
Cool.

I couldn't figure out how to get flake8 to tell me what it was doing, but if I have time I'll try your suggestion. Meanwhile, the repo I was testing against is [atomic-reactor](https://github.com/containerbuildsystem/atomic-reactor), right here on GitHub :)

**Comment by sondrelg** [comment]
Looks like I found something 👍 

After pulling atomic-reactor and running just flake8 + this package, it runs with no errors. But after installing `flake8-bandit` and running it again, I get the same error as you! 🥳 

I'll see if I can narrow it down myself for now and report back 🙂

**Comment by sondrelg** [comment]
Think I've got to the bottom of it 🥳 

So the ast objects we receive form flake8 to our plugin, include the ast-objects from the source code we're checking, but can also sometimes include elements left over from other plugins. 

The dict representation of a node might look like this:

```python
{
    # What we care about
    'value': <_ast.Name object at 0x10e4241f0>, 
    'attr': 'key', 'ctx': <_ast.Load object at 0x10d469bb0>, 
    'lineno': 35, 'col_offset': 20, 
    'end_lineno': 35, 
    'end_col_offset': 40, 
    # Plugin poop
    '_pyflakes_depth': 5, 
    '_pyflakes_parent': <_ast.Dict object at 0x10e424130>, 
    '_bandit_sibling': <_ast.Dict object at 0x10e424220>, 
    '_bandit_parent': <_ast.Dict object at 0x10e424130>
}
```

There are no `None` values to consider in the Python ast objects we receive, but it seems that `flake8-bandit` sometimes leave the attribute `_bandit_sibling` as `None`, so the combination of the two packages caused the error 🕵️ 

This should do the trick - pretty much what you suggested above:

```diff
    for key, value in node.__dict__.items():
-        if type(value) not in [int, str, list, bool]:
+        if type(value) not in [int, str, list, bool] and value is not None:
            setattr(node.__dict__[key], attr, val)
```

I'll push a new version shortly 🏄‍♂️

**Comment by sondrelg** [comment]
If you run `pip install flake8-type-checking -U`, I believe your problem should be resolved 👏  Just pushed the fix

Please let me know if it works @ben-alkov 🙂

**Comment by ben-alkov** [comment]
Drat it, lost my comment.

tldr; As shipped, it still throws a TB for me. I think a good solution here is what I was trying to figure out this morning (but I'd forgotten the finer points of `isinstance()`; thanks again, [S.O.](https://stackoverflow.com/questions/21706609/where-is-the-nonetype-located-in-python-3-x)!) - to wit: test with `isinstance(value, (int, str, list, bool, str, type(None)))`

**Comment by sondrelg** [comment]
Ok good to know 🙂  Just to confirm, are you running `v1.0.1` and it's still exactly the same error as well? 

Would you be interested in trying to create a fix? Contributions are always welcome 👏   

No problem if not; I'll jump back onto it tomorrow most likely if so 👍

**Comment by sondrelg** [comment]
It just hit me that I didn't actually implement my own fix yesterday 🤦‍♂️  I blame lack of sleep

Could you test one more time with v1.0.2 @ben-alkov?

**Comment by ben-alkov** [comment]
Fixed in 1.0.2!

**Comment by sondrelg** [comment]
Wonderful 👏 

Thanks @ben-alkov! Let me know if you find anything else 😄
